In [0]:
df1=spark.read.load("/Volumes/luffy/phase2/silver/df1/")
display(df1)

CustomerID total_spent total_transactions total_quantity last_purchase_date is_high_valued class_weight 17420 598.8300000000002 30 265 2011-10-20 0 0.6302969155376189 16552 379.72999999999996 17 219 2011-04-11 0 0.6302969155376189 17572 226.75 12 95 2011-09-29 0 0.6302969155376189 15350 115.65 5 51 2010-12-01 0 0.6302969155376189 12921 16784.440000000013 741 9598 2011-12-06 1 2.418694690265487 13090 9132.249999999998 161 2332 2011-12-01 1 2.418694690265487 14135 4690.31 134 3850 2011-12-08 1 2.418694690265487 12915 363.65 22 93 2011-07-14 0 0.6302969155376189 17685 3191.5299999999993 130 1956 2011-11-23 1 2.418694690265487 17581 11353.969999999987 452 5993 2011-12-09 1 2.418694690265487 14031 19939.080000000005 258 6206 2011-12-02 1 2.418694690265487 16995 1.25 1 1 2010-12-02 0 0.6302969155376189 17961 1296.4400000000012 249 560 2011-11-18 0 0.6302969155376189 13370 788.4200000000004 44 350 2010-12-08 0 0.6302969155376189 13013 4834.89 222 2231 2011-12-08 1 2.418694690265487 16926 230.25000000000003 14 35 2010-12-03 0 0.6302969155376189 12725 427.79999999999995 24 164 2010-12-03 0 0.6302969155376189 16725 786.9700000000004 152 295 2011-02-10 0 0.6302969155376189 16402 299.59999999999997 60 149 2011-03-20 0 0.6302969155376189 13652 1095.08 63 578 2011-10-25 0 0.6302969155376189 16858 405.49 13 99 2010-12-08 0 0.6302969155376189 18113 76.32000000000001 1 72 2010-12-06 0 0.6302969155376189 15021 1957.3600000000029 479 865 2011-12-01 0 0.6302969155376189 13319 11026.759999999986 500 4762 2011-12-04 1 2.418694690265487 16306 662.2700000000001 44 503 2011-05-10 0 0.6302969155376189 12839 5599.220000000003 316 3810 2011-12-07 1 2.418694690265487 17975 4584.010000000002 297 1584 2011-11-24 1 2.418694690265487 15973 515.87 6 125 2011-04-18 0 0.6302969155376189 18061 2119.4099999999994 46 955 2011-11-18 1 2.418694690265487 18259 2338.600000000001 42 714 2011-11-15 1 2.418694690265487 12963 1872.2299999999996 96 889 2011-12-01 0 0.6302969155376189 17671 1861.3400000000004 61 885 2011-11-24 0 0.6302969155376189 16579 30.599999999999998 1 12 2010-12-09 0 0.6302969155376189 17442 2286.86 140 1502 2011-11-10 1 2.418694690265487 13969 9121.190000000002 633 4601 2011-12-02 1 2.418694690265487 14419 1058.0600000000002 206 487 2011-11-29 0 0.6302969155376189 16019 4122.239999999999 165 1817 2011-12-08 1 2.418694690265487 15044 7771.079999999996 461 2804 2011-11-20 1 2.418694690265487 17303 250.03999999999994 53 144 2010-12-13 0 0.6302969155376189 16503 1442.4300000000005 86 571 2011-08-25 0 0.6302969155376189 13238 720.6300000000001 32 300 2011-11-28 0 0.6302969155376189 14401 4062.9799999999987 183 4509 2011-12-05 1 2.418694690265487 17525 422.70000000000005 9 82 2010-12-15 0 0.6302969155376189 13458 9755.22 394 5556 2011-12-02 1 2.418694690265487 17551 306.84 43 150 2010-12-15 0 0.6302969155376189 17752 80.64 1 192 2010-12-15 0 0.6302969155376189 15246 514.1500000000001 33 247 2011-04-15 0 0.6302969155376189 15759 253.04000000000002 7 60 2011-05-05 0 0.6302969155376189 16684 67415.04000000001 281 51120 2011-12-05 1 2.418694690265487 17848 901.2000000000003 51 398 2011-09-19 0 0.6302969155376189 16153 1824.2300000000002 89 1073 2011-10-21 0 0.6302969155376189 16282 168.3 11 114 2011-01-04 0 0.6302969155376189 17389 32367.280000000013 224 7794 2011-12-09 1 2.418694690265487 15046 5363.530000000002 207 3595 2011-11-30 1 2.418694690265487 17128 157.09 14 76 2011-01-09 0 0.6302969155376189 12651 114.00000000000003 9 78 2011-01-09 0 0.6302969155376189 17800 1199.01 104 736 2011-11-02 0 0.6302969155376189 12652 925.4299999999998 51 508 2011-01-21 0 0.6302969155376189 15447 155.17 9 85 2011-01-13 0 0.6302969155376189 12490 5417.9299999999985 245 3064 2011-12-04 1 2.418694690265487 13048 1865.9099999999994 137 1246 2011-11-25 0 0.6302969155376189 14657 858.1700000000001 127 467 2011-11-22 0 0.6302969155376189 18233 440.0 1 4 2011-01-18 0 0.6302969155376189 12451 9137.519999999997 355 6695 2011-11-29 1 2.418694690265487 14784 289.99 15 237 2011-11-13 0 

In [0]:
train_df, test_df = df1.randomSplit([0.8, 0.2], seed=42)
display(train_df)
display(test_df)

CustomerID total_spent total_transactions total_quantity last_purchase_date is_high_valued class_weight null 2062871.1599989405 135080 689008 2011-12-09 1 2.418694690265487 12346 154367.2 2 148430 2011-01-18 1 2.418694690265487 12347 4309.999999999997 182 2458 2011-12-07 1 2.418694690265487 12348 1797.24 31 2341 2011-09-25 0 0.6302969155376189 12349 1757.55 73 631 2011-11-21 0 0.6302969155376189 12350 334.40000000000003 17 197 2011-02-02 0 0.6302969155376189 12352 3466.6700000000014 95 602 2011-11-03 1 2.418694690265487 12353 89.0 4 20 2011-05-19 0 0.6302969155376189 12355 459.4 13 240 2011-05-09 0 0.6302969155376189 12356 2811.4300000000007 59 1591 2011-11-17 1 2.418694690265487 12358 1168.06 19 248 2011-12-08 0 0.6302969155376189 12360 2662.0600000000004 129 1165 2011-10-18 1 2.418694690265487 12361 189.89999999999998 10 91 2011-02-25 0 0.6302969155376189 12362 5297.880000000003 274 2246 2011-12-06 1 2.418694690265487 12363 552.0 23 408 2011-08-22 0 0.6302969155376189 12364 1313.1000000000001 85 1506 2011-12-02 0 0.6302969155376189 12365 962.0699999999999 23 175 2011-02-21 0 0.6302969155376189 12367 168.9 11 173 2011-12-05 0 0.6302969155376189 12371 1887.9600000000003 63 591 2011-10-26 0 0.6302969155376189 12375 459.5800000000001 18 179 2011-12-07 0 0.6302969155376189 12378 4008.6200000000013 219 2529 2011-08-02 1 2.418694690265487 12379 854.1899999999997 41 408 2011-09-19 0 0.6302969155376189 12381 1886.6600000000008 91 833 2011-12-05 0 0.6302969155376189 12383 1861.8099999999997 100 1535 2011-06-08 0 0.6302969155376189 12384 604.38 30 131 2011-11-11 0 0.6302969155376189 12388 2780.6600000000003 100 1462 2011-11-24 1 2.418694690265487 12390 549.8399999999999 32 353 2011-09-21 0 0.6302969155376189 12391 460.88999999999993 100 300 2011-11-18 0 0.6302969155376189 12393 1582.6000000000004 64 816 2011-09-28 0 0.6302969155376189 12394 1272.48 27 819 2011-10-07 0 0.6302969155376189 12397 2409.9000000000005 126 1263 2011-11-04 1 2.418694690265487 12398 1635.6599999999999 85 1394 2011-10-25 0 0.6302969155376189 12399 1108.65 59 1178 2011-08-12 0 0.6302969155376189 12401 84.3 5 11 2011-02-09 0 0.6302969155376189 12402 225.6 11 94 2011-01-20 0 0.6302969155376189 12406 3549.8500000000013 108 1827 2011-11-17 1 2.418694690265487 12407 1708.1199999999997 77 1416 2011-10-21 0 0.6302969155376189 12409 11088.41 114 5559 2011-09-22 1 2.418694690265487 12410 705.5799999999999 28 213 2011-02-11 0 0.6302969155376189 12412 1226.8799999999999 61 829 2011-09-26 0 0.6302969155376189 12414 562.41 18 300 2011-05-06 0 0.6302969155376189 12415 126103.60999999964 778 78098 2011-11-15 1 2.418694690265487 12417 3719.3999999999996 198 1565 2011-12-06 1 2.418694690265487 12418 668.5799999999997 34 407 2011-08-19 0 0.6302969155376189 12420 600.3900000000001 29 345 2011-10-07 0 0.6302969155376189 12421 807.8100000000002 46 485 2011-11-24 0 0.6302969155376189 12422 803.5599999999998 39 370 2011-09-05 0 0.6302969155376189 12424 1760.9599999999998 30 740 2011-06-30 0 0.6302969155376189 12425 777.1999999999999 36 496 2011-09-22 0 0.6302969155376189 12426 582.7299999999999 30 258 2011-05-29 0 0.6302969155376189 12427 943.2300000000002 35 830 2011-11-28 0 0.6302969155376189 12429 3750.400000000001 97 1557 2011-11-30 1 2.418694690265487 12430 256.4 11 117 2011-10-27 0 0.6302969155376189 12431 6558.509999999998 240 2968 2011-11-04 1 2.418694690265487 12432 5059.3200000000015 108 1801 2011-10-28 1 2.418694690265487 12433 13375.869999999992 420 11071 2011-12-09 1 2.418694690265487 12435 7829.889999999998 36 2055 2011-09-21 1 2.418694690265487 12436 509.01000000000005 16 211 2011-09-01 0 0.6302969155376189 12437 5006.159999999997 201 2928 2011-12-08 1 2.418694690265487 12438 2906.8500000000004 98 1180 2011-11-25 1 2.418694690265487 12441 173.55 11 121 2010-12-08 0 0.6302969155376189 12442 172.06 12 182 2011-12-06 0 0.6302969155376189 12444 5005.46 177 1155 2011-11-18 1 2.418694690265487 12447 476.49 26 190 2011-04-10 0 0.6302969155376189 12448 449.4500000000001 22 243 20

CustomerID total_spent total_transactions total_quantity last_purchase_date is_high_valued class_weight 12354 1079.4 58 530 2011-04-21 0 0.6302969155376189 12357 6207.669999999996 131 2708 2011-11-06 1 2.418694690265487 12359 6499.630000000005 254 1632 2011-12-02 1 2.418694690265487 12370 3545.689999999998 167 2353 2011-10-19 1 2.418694690265487 12372 1298.0400000000002 52 794 2011-09-29 0 0.6302969155376189 12373 364.6 14 197 2011-02-01 0 0.6302969155376189 12374 742.93 33 342 2011-11-14 0 0.6302969155376189 12377 1628.1199999999997 77 944 2011-01-28 0 0.6302969155376189 12380 2729.06 105 1128 2011-11-18 1 2.418694690265487 12386 401.90000000000003 10 354 2011-01-06 0 0.6302969155376189 12395 3038.9799999999987 159 2123 2011-11-24 1 2.418694690265487 12403 445.7 6 97 2011-10-21 0 0.6302969155376189 12405 1710.3900000000006 54 857 2011-07-14 0 0.6302969155376189 12408 2934.530000000001 113 1419 2011-11-07 1 2.418694690265487 12413 822.05 40 290 2011-10-04 0 0.6302969155376189 12423 1869.5100000000002 126 1351 2011-12-09 0 0.6302969155376189 12428 8035.720000000003 300 3569 2011-11-14 1 2.418694690265487 12434 847.0899999999999 54 415 2011-09-16 0 0.6302969155376189 12445 133.4 4 62 2011-11-17 0 0.6302969155376189 12446 1002.3099999999998 58 352 2011-10-13 0 0.6302969155376189 12449 4067.2900000000013 191 1553 2011-11-17 1 2.418694690265487 12451 9137.519999999997 355 6695 2011-11-29 1 2.418694690265487 12457 2808.6799999999994 68 993 2011-10-12 1 2.418694690265487 12480 3281.6300000000015 154 1707 2011-11-11 1 2.418694690265487 12493 450.69 23 208 2011-06-27 0 0.6302969155376189 12502 3723.87 147 820 2011-09-05 1 2.418694690265487 12519 340.84 18 183 2011-10-07 0 0.6302969155376189 12521 599.6800000000001 38 309 2011-07-20 0 0.6302969155376189 12523 1833.9700000000007 110 869 2011-12-08 0 0.6302969155376189 12538 363.95 21 191 2011-11-10 0 0.6302969155376189 12548 95.19999999999999 5 49 2011-06-26 0 0.6302969155376189 12551 168.0 10 100 2010-12-17 0 0.6302969155376189 12557 11990.96 33 4384 2011-11-23 1 2.418694690265487 12569 4184.3899999999985 191 1809 2011-12-07 1 2.418694690265487 12571 519.6099999999999 34 523 2011-10-31 0 0.6302969155376189 12581 197.70999999999998 15 161 2011-10-31 0 0.6302969155376189 12588 214.9 10 53 2011-10-31 0 0.6302969155376189 12593 418.09999999999997 21 206 2011-08-23 0 0.6302969155376189 12600 2617.0400000000004 65 775 2011-11-11 1 2.418694690265487 12604 254.17999999999995 16 201 2011-09-21 0 0.6302969155376189 12609 1126.8500000000001 60 648 2011-09-22 0 0.6302969155376189 12611 1193.32 65 846 2011-10-18 0 0.6302969155376189 12616 769.3499999999999 36 295 2011-09-15 0 0.6302969155376189 12617 355.5300000000002 42 213 2011-11-10 0 0.6302969155376189 12620 1339.4099999999999 92 1181 2011-11-25 0 0.6302969155376189 12625 2753.3199999999993 91 1647 2011-05-12 1 2.418694690265487 12626 6852.609999999996 307 3556 2011-11-16 1 2.418694690265487 12628 803.3099999999998 43 451 2011-12-05 0 0.6302969155376189 12631 640.5000000000002 31 236 2011-10-13 0 0.6302969155376189 12639 486.1 22 201 2011-04-15 0 0.6302969155376189 12643 3344.6400000000003 24 1808 2011-08-03 1 2.418694690265487 12648 318.05 21 160 2011-07-08 0 0.6302969155376189 12649 643.31 33 598 2011-08-24 0 0.6302969155376189 12660 245.10000000000002 13 127 2011-11-28 0 0.6302969155376189 12666 227.44 2 56 2010-12-15 0 0.6302969155376189 12680 862.81 52 443 2011-12-09 0 0.6302969155376189 12689 1247.61 72 628 2011-11-14 0 0.6302969155376189 12693 880.86 41 260 2011-08-11 0 0.6302969155376189 12695 582.1 27 220 2011-12-02 0 0.6302969155376189 12701 797.7099999999999 53 678 2011-09-21 0 0.6302969155376189 12709 10178.939999999988 257 4040 2011-12-06 1 2.418694690265487 12712 4550.63 142 1080 2011-11-17 1 2.418694690265487 12713 848.5499999999998 38 508 2011-12-09 0 0.6302969155376189 12719 681.7999999999998 42 387 2011-12-04 0 0.6302969155376189 12721 2683.280000000001 164 1964 2011-11-08 1 2.418694690265487 12728 836.3299999999999 46 419 2011

In [0]:
train_df.write.format("delta").mode("overwrite").save("/Volumes/luffy/phase2/silver/train_df")
test_df.write.format("delta").mode("overwrite").save("/Volumes/luffy/phase2/silver/test_df")

In [0]:
from pyspark.ml.feature import VectorAssembler

assembler = VectorAssembler(
    inputCols=["total_spent", "total_transactions", "total_quantity"],
    outputCol="features"
)

train_df = assembler.transform(train_df)
test_df = assembler.transform(test_df)

# Training the Logistic Regression Model

In [0]:
from pyspark.ml.classification import LogisticRegression

lr = LogisticRegression(
    featuresCol="features",
    labelCol="is_high_valued",
    weightCol="class_weight"
)

lr_model = lr.fit(train_df)

# Testing the Model

In [0]:
lr_predictions = lr_model.transform(test_df)
from pyspark.ml.evaluation import BinaryClassificationEvaluator

evaluator = BinaryClassificationEvaluator(
    labelCol="is_high_valued",
    rawPredictionCol="rawPrediction",
    metricName="areaUnderROC"
)

lr_auc = evaluator.evaluate(lr_predictions)
print("Logistic Regression AUC:", lr_auc)

Logistic Regression AUC: 1.0


# Training RandomForest Model

In [0]:
from pyspark.ml.classification import RandomForestClassifier

rf = RandomForestClassifier(
    featuresCol="features",
    labelCol="is_high_valued",
    weightCol="class_weight",
    numTrees=27
)

rf_model = rf.fit(train_df)
rf_predictions = rf_model.transform(test_df)

# Testing

In [0]:
rf_auc = evaluator.evaluate(rf_predictions)
print("RandomForest AUC:", rf_auc)

RandomForest AUC: 0.9988120596427986


# Comparing AUC
We got AUC = 1 for Logistic Regression and AUC = 0.9988120596427986 for RandomForest